plugins/UsersManager/tests/Integration/TokenPostOnlyTest.php

```diff
@@ -0,0 +1,88 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+namespace Piwik\Plugins\UsersManager\tests\Integration;
+
+use Piwik\Plugins\UsersManager\API as UsersManagerAPI;
+use Piwik\Plugins\UsersManager\Model as UsersManagerModel;
+use Piwik\Tests\Framework\Fixture;
+use Piwik\Tests\Framework\TestCase\IntegrationTestCase;
+
+/**
+ * @group UsersManager
+ * @group TokenPostOnlyTest
+ */
+class TokenPostOnlyTest extends IntegrationTestCase
+{
+    protected static $tokenPostOnly = 'f3fa8c38fd277a9af0fab7e35f9736fe';
+
+    public function setUp(): void
+    {
+        parent::setUp();
+        self::createUserAndTokens();
+    }
+
+    private static function createUserAndTokens()
+    {
+        if (!Fixture::siteCreated($idSite = 1)) {
+            Fixture::createWebsite('2021-01-01');
+        }
+
+        if (!UsersManagerAPI::getInstance()->userExists('user1')) {
+            UsersManagerAPI::getInstance()->addUser('user1', 'user1password', 'user@limited.com');
+            UsersManagerAPI::getInstance()->setUserAccess('user1', 'view', [1]);
+
+            $userModel = new UsersManagerModel();
+            $userModel->addTokenAuth('user1', self::$tokenPostOnly, 'Post Only', '2020-01-02 03:04:05',
+                null, false, true);
+        }
+    }
+
+    /**
+     * Post Only tokens should return a 401 code if used in a GET request
+     */
+    public function test_postOnlyToken_AccessDeniedIfGet()
+    {
+        $url = Fixture::getTestRootUrl().'?'.http_build_query([
+                'module'     => 'API',
+                'method'     => 'API.getMatomoVersion',
+                'token_auth' => self::$tokenPostOnly,
+            ]);
+
+        $ch = curl_init();
+        curl_setopt($ch, CURLOPT_URL, $url);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+        $out = curl_exec($ch);
+        $responseInfo = curl_getinfo($ch);
+        curl_close($ch);
+
+        $this->assertEquals(401, $responseInfo["http_code"]);
+        $this->assertStringContainsString("or is required to be sent as a POST parameter", $out);
+    }
+
+    /**
+     * Post only tokens should return a 200 code if used in a POST request
+     */
+    public function test_postOnlyToken_AccessGrantedIfPost()
+    {
+        $url = Fixture::getTestRootUrl().'?'.http_build_query([
+                'module'     => 'API',
+                'method'     => 'API.getMatomoVersion'
+            ]);
+
+        $ch = curl_init();
+        curl_setopt($ch, CURLOPT_POSTFIELDS, ['token_auth' => self::$tokenPostOnly]);
+        curl_setopt($ch, CURLOPT_URL, $url);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+        curl_exec($ch);
+        $responseInfo = curl_getinfo($ch);
+        curl_close($ch);
+
+        $this->assertEquals(200, $responseInfo["http_code"]);
+    }
```

Comment:
@sgiehl Could you point me to a similar test for session a token_auth? Not quite sure what's required here.

Reply:
I don't think we currently have any test doing something similar. Might be something like faking a session login so a session token auth is generated, which should then be valid. But thinking about this, this might be somewhat covered by ui tests.
Btw. there are currently a couple of UI tests failing as the token is not valid. Seems something needs to be adjusted there...

Reply:
@sgiehl I've added a fix for the failing UI tests, the new exceptions should now only be thrown for API requests. I'm the `Login->isModuleIsAPI()` method to determine this, not sure if there is a better option?

Reply:
I guess it should be fine that way. Will have another look at this PR tomorrow.